### PR TITLE
feat(inputs.procstat): Allow multiple selection criteria

### LINKED ISSUE #6174: Procstat should support multiple search criteria
### Relevant telegraf.conf:

```
[[inputs.procstat]]
  exe = "java"
  user = "logstash|elasticsearch"
  interval = "1m"
  pid_tag = true

[[inputs.procstat]]
  exe = "influxd"
  user = "influxdb"
  interval = "1m"
  pid_tag = false

[[inputs.procstat]]
  exe = "node"
  user = "kibana"
  interval = "1m"
```

### System info:

telegraf-1.11.1-1.x86_64
influxdb-1.7.7-1.x86_64

CentOS Linux release 7.6.1810 (Core)

### Steps to reproduce:
the telegraf github page show way more fields and tags for the procstat_lookup measurement than we can see. tried to change pid_tag = true/false, no success

### Expected behavior:
we need at least the 'user' tag in order to differentiate between the 2 java processes (elasticsearch+logstash)
```
procstat_lookup
tags:
- exe
- pid_finder
- pid_file
- pattern
- prefix
- user
- systemd_unit
- cgroup
- win_service
- result

fields:
- pid_count (int)
- running (int)
- result_code (int, success = 0, lookup_error = 1)
```
### Actual behavior:
```
> select * from procstat_lookup
name: procstat_lookup
time                 exe     host               pid_count pid_finder result  result_code running
----                 ---     ----               --------- ---------- ------  ----------- -------
2019-07-26T08:26:00Z influxd MYHOSTNAME 1         pgrep      success 0           1
2019-07-26T08:26:00Z java    MYHOSTNAME 2         pgrep      success 0           2
2019-07-26T08:26:00Z node    MYHOSTNAME 1         pgrep      success 0           1
```

### PR DESCRIPTION
**Pull request by srebhan**
## Summary

This PR adds the possibility to specify multiple filter expressions for the process list. Each filter is independent of the other ones, intra-criterion lists are combined using logical **or** whereas different criteria are combined via logical **and**.

```toml
[[inputs.procstat]]
  [[inputs.procstat.filter]]
    name = "shells"
    patterns = ['fish']
    users = ["sv*"]
```

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #6174

### CODE REVIEW

**Comment by tguenneguez** [comment]
He,

Haven't test it, but It seems correct overall.

I doesn't see how the "pgrep" mode work ?
I think this new implementation will only be available with "native" mode ? Why have a double code ?

The Filter should have more criterias :
exe (the name of binary (/proc/*/exe)), see https://pkg.go.dev/github.com/shirou/gopsutil/v3/process#Process.Exe
name (content of process (/proc/*/comm)), see https://pkg.go.dev/github.com/shirou/gopsutil/v3/process#Process.Name

It's difficult because de multifilter is only available for :
name, patterns and users
In your code it's also possible with other filter, but it doesn't should work ?

It's strange to agregate multi filter values in the procstat metrics ?
If I understand your implementation with this conf :
[[inputs.procstat]]
  [[inputs.procstat.filter]]
    name = "crond"
  [[inputs.procstat.filter]]
    name = "sshd"
You will have procstat metrics wich agregate metrics of 2 types off processus. It's strange ?

Thanks
Thomas

**Comment by srebhan** [comment]
@tguenneguez I would love to get some help from you to implement the missing filters. This was meant as a proof-of-concept and is not finished yet... So please feel free to push to my PR!

Regarding `pgrep`, I'm not sure it even makes sense to add this here. The `processes` package of gopsutil is available for all supported platforms AFAICS so what is the benefit for doing it in two different ways?

**Comment by tguenneguez** [comment]
> @tguenneguez I would love to get some help from you to implement the missing filters. This was meant as a proof-of-concept and is not finished yet... So please feel free to push to my PR!
> 
> Regarding `pgrep`, I'm not sure it even makes sense to add this here. The `processes` package of gopsutil is available for all supported platforms AFAICS so what is the benefit for doing it in two different ways?

100% agree with you.
I will see for help you to imprement good feature

**Comment by tguenneguez** [comment]
Before complete your job, there are things to clarify.

The "name" should be add by [inputs.procstat.tags].
I think it's not a good idea to add this feature in your code...

I also think you will have in one part :
pid_files, systemd_units, cgroups, supervisor_units, win_services
And in the other hand, filters.

Indeed, the first filters are already very restrictive and I don't think there is any point in combining them. Besides, the implementation does use "pid_finder".
So I would leave these first criteria aside.

In filter, I would only implement the filters:
name
exe
pattern (that will be renamed in "cmdLine")
user

Name is the content of /proc/<PID>/comm
Exe is the link of /proc/<PID>/exe

In Filters, it seem's to define some filters as a liste...
As the filters are cumulative, I don't think it makes sense to be able to put lists.
Wath is the meaning of :
[[inputs.procstat.filter]]
  pattern = ["myproc", "system"]
  users = ["root", "cft"]
Should it return processes that match all the criteria, one of the criteria, a pattern and a user, ...?

looking forward to reading you
Thomas

**Comment by srebhan** [comment]
Thanks @tguenneguez for your comments! I'm trying to answer one-by-one:

> The "name" should be add by [inputs.procstat.tags].
I think it's not a good idea to add this feature in your code...

The user has to have some ability to find out which of the potentially multiple filters produced the data, so I think we need to add the filter name. If you don't like `filter` for the tag name, I'm open for suggestions, but you cannot do this via `[inputs.procstat.tags]` as those are the same for all filters.

> I also think you will have in one part :
pid_files, systemd_units, cgroups, supervisor_units, win_services
And in the other hand, filters.

Why do you want to add this limitation? Imagine you want to monitor 4 different systemd units but only certain child processes or different users. In your scheme, this means four different plugin instances for no good reason...

 > Indeed, the first filters are already very restrictive and I don't think there is any point in combining them. Besides, the implementation does use "pid_finder".
So I would leave these first criteria aside.

This might be true for your use-cases but I know that people are looking for combining those to monitor only partial aspects...

> [...] Name is the content of /proc//comm
Exe is the link of /proc//exe

I'm not against having name and exe, but please also think about other operating systems such as Windows or MacOS.

> In Filters, it seem's to define some filters as a liste...
As the filters are cumulative, I don't think it makes sense to be able to put lists.
Wath is the meaning of :
[[inputs.procstat.filter]]
pattern = ["myproc", "system"]
users = ["root", "cft"]
Should it return processes that match all the criteria, one of the criteria, a pattern and a user, ...?

In fact, all filters are lists! Each entry, i.e. `pattern` or `users` is a **or** concatenated list while the entries are combined via **and**. So your example means the process matches `(pattern == "myproc" || pattern == "system") && (user == "root" || user == "cft"))`. Feel free to improve the documentation!

I hope this clarifies some points!?

**Comment by powersj** [comment]
@tguenneguez have you had a chance to look over @srebhan's response above? Were you going to update your existing PR based on this?

**Comment by tguenneguez** [comment]
> @tguenneguez have you had a chance to look over @srebhan's response above? Were you going to update your existing PR based on this?

I will see when I will have time but not for the moment sorry

File: plugins/inputs/procstat/README.md [hunk 1]
```diff
@@ -77,6 +77,34 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
...
+  #    # patterns = ['.*']
+  #    ## List of users owning the process (wildcards are supported)
+  #    # users = ['*']
+  #    ## List of executable paths of the process (wildcards are supported)
+  #    # executables = ['*']
+  #    ## List of process names (wildcards are supported)
+  #    # process_names = ['*']
+  #    ## Recursion depth for determining children of the matched processes
+  #    ## A negative value means all children with infinite depth
+  #    #recursion_depth = 0
```
Comment (powersj) [inline plugins/inputs/procstat/README.md:107]:
```suggestion
  #    # recursion_depth = 0
```

File: plugins/inputs/procstat/sample.conf [hunk 2]
```diff
@@ -48,3 +48,31 @@
...
+  #    # patterns = ['.*']
+  #    ## List of users owning the process (wildcards are supported)
+  #    # users = ['*']
+  #    ## List of executable paths of the process (wildcards are supported)
+  #    # executables = ['*']
+  #    ## List of process names (wildcards are supported)
+  #    # process_names = ['*']
+  #    ## Recursion depth for determining children of the matched processes
+  #    ## A negative value means all children with infinite depth
+  #    #recursion_depth = 0
```
Comment (powersj) [inline plugins/inputs/procstat/sample.conf:78]:
```suggestion
  #    # recursion_depth = 0
```

File: plugins/inputs/procstat/sample.conf [hunk 3]
```diff
@@ -48,3 +48,31 @@
...
+  #    ## Systemd unit names (wildcards are supported)
+  #    # systemd_units = []
+  #    ## CGroup name or path (wildcards are supported)
+  #    # cgroups = []
+  #    ## Supervisor service names of hypervisorctl management
+  #    # supervisor_units = []
+  #    ## Windows service names
+  #    # win_service = []
+  #
+  #    ## Process filters, multiple ones are allowed
```
Comment (powersj) [inline plugins/inputs/procstat/sample.conf:67]:
```suggestion
  #    ## Process filters, multiple are allowed
```

File: plugins/inputs/procstat/README.md [hunk 4]
```diff
@@ -77,6 +77,34 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
...
+  #    ## Systemd unit names (wildcards are supported)
+  #    # systemd_units = []
+  #    ## CGroup name or path (wildcards are supported)
+  #    # cgroups = []
+  #    ## Supervisor service names of hypervisorctl management
+  #    # supervisor_units = []
+  #    ## Windows service names
+  #    # win_service = []
+  #
+  #    ## Process filters, multiple ones are allowed
```
Comment (powersj) [inline plugins/inputs/procstat/README.md:96]:
```suggestion
  #    ## Process filters, multiple are allowed
```

**Review by powersj — APPROVED** [review]
Looks good, the new `findByPidFiles` where you verify if the PID exists would be good for #15017.

Made a couple small changes to sample conf & readme.

**Review by DStrand1 — APPROVED** [review]
LGTM! Will leave to you to merge after resolving the review comments

**Pull request merged**